libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/extensions/WaypointEx.kt

```diff
@@ -0,0 +1,41 @@
+@file:JvmName("WaypointEx")
+
+package com.mapbox.navigation.base.internal.extensions
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.internal.route.Waypoint
+import com.mapbox.navigation.base.trip.model.RouteProgress
+
+/**
+ * Return waypoints that are requested explicitly
+ */
+fun List<Waypoint>.requestedWaypoints(): List<Waypoint> =
+    this.filter { it.type != Waypoint.EV_CHARGING }
+
+/**
+ * Return waypoints that tracking in [RouteProgress.currentLegProgress]#legIndex and based on
+ * [DirectionsRoute.legs] index
+ */
+fun List<Waypoint>.legsWaypoints(): List<Waypoint> =
+    this.filter { it.type != Waypoint.SILENT }
+
+/**
+ * Return the index of **next requested** coordinate.
+ *
+ * For instance, EV waypoints are not requested explicitly, so they are not taken into account.
+ */
+fun indexOfNextRequestedCoordinate(
+    waypoints: List<Waypoint>,
+    remainingWaypoints: Int,
+): Int? {
+    if (remainingWaypoints > waypoints.size) {
+        return null
+    }
+    val nextRequestedWaypoint = waypoints
+        .subList(waypoints.size - remainingWaypoints, waypoints.size)
+        .requestedWaypoints()
+        .firstOrNull()
+        ?: return null
+
+    return waypoints.indexOf(nextRequestedWaypoint)
```

Comment:
Please correct me if I'm wrong but it looks like the index returned here is the index in the `Waypoint` list which contains EV waypoints but we use this index to mutate the original request which does not cotain for EV waypoints. I think we're looking for something like this instead:
```suggestion
    return waypoints.requestedWaypoints().indexOf(nextRequestedWaypoint)
```

Reply:
I think this method is about finding the right index in the passed list.

Reply:
you're both right 😄  I tried to avoid calling the second `requestedWaypoints` but missed that it's required to find the right index (tests become 🔴, so we're in safety 🥲  ). Already in the branch